### DISCUSSION TITLE
Binary compatibility on importing COM dll/exe projects

### DISCUSSION THREAD

**Discussion by dude202** [twinbasic/twinbasic#1568] [General]
When do you plan to have it, is it possible (will it be any time soon) to manually set all relevant GUIDs after the import to achieve binary compatibility that way? Thanks.

**Comment by fafalone** [comment]
You can already set all the GUIDs afaik but isn't there more to binary compatibility?

**Reply by dude202** [reply]
Order of public methods? DispIds? (can they be defined in TwinBasic?) Basically, there are two questions - 1. can TwinBasic (as of now) at all define everything needed in code and 2. will IDE import process (sometime soon) help setting all those automatically based on a given compatibility dll, as VB does? The first question is crucial for transitioning existing VB to TwinBasic, the second is just a nice helper that saves a lot of time.

**Reply by fafalone** [reply]
AFAIK methods are ordered the same as VB6-- as they appear, or as defined by the interface (there's a pending issue with overloads; it breaks COM rules so is technically undefined behavior-- MSVC compiler reverses them, tB currently does not). 

Yes, you can define DispIDs with the `[ DispId(#) ]` attribute, e.g.

```
[ InterfaceId ("2D91EEA1-9932-11d2-BE86-00A0C9A83DA1") ]
[ OleAutomation (True) ]
Interface IFileSearchBand Extends IDispatch
    [ DispId (1) ] Sub SetFocus()
    [ DispId (2) ] Sub SetSearchParameters(ByVal pbstrSearch As String, ByVal bNavToResults As Boolean, Optional pvarScope As Variant, Optional pvarQueryFile As Variant)
    [ DispId (3) ] Property Get SearchID() As String
    [ DispId (4) ] Property Get Scope() As Variant
    [ DispId (5) ] Property Get QueryFile() As Variant
End Interface
```

I'm not sure VB6 files expose the GUIDs and such in the saved files to import. But tB has a typelib viewer that you could copy/paste from.

For UCs, they're headed by 

```[ FormDesignerId ("8CE9DEA3-CF96-4344-87A8-3F6F9ECBBD42") ]
[ ClassId ("64B7FBF6-FCB5-470D-A624-EB45FDD4A7A0") ]
[ InterfaceId ("D656BF4B-2A6B-4835-B172-F646811CD3E9") ]
[ EventInterfaceId ("0F271F22-C9E5-4128-AC76-0C816C54B2CB") ]
[ COMControl ]
Public Class ucShellBrowse
```
for example, with a Project GUID/TypeLib GUID defined in Project Settings.
